### ISSUE TITLE
Auto-approve a fraction of WebExtension updates

### ISSUE THREAD

**Issue by jvillalobos** [mozilla/addons-server#4533]
(Edited to keep up with proposed changes)

We're currently experiencing significant delays in reviews due to decreased reviewer activity and increase submissions. It looks like this will only get worse leading to the transition to WebExtensions, so we need to start moving to a place where we automatically review at least some add-ons.

This is a first step that should hopefully be easy to implement quickly. For more context, please refer to [this draft PRD](https://docs.google.com/document/d/1VxvvzQQvfDOVCylSVlpfeXv0tDYRLZfIMvWOoraANGU/edit#).

Add-on submissions will be auto-approved based on a submission counter. In order to qualify for auto-approval, the add-ons need to meet this criteria:

1. Have been manually approved on initial submission.
2. Are not flagged for admin review. This can happen automatically on submission in some cases, but is generally done manually by reviewers.
3. WebExtensions (excludes Embedded WebExtensions).
4. Active Daily Users under a threshold. Initially 10K, but configurable.
5. Don’t trigger any of the following linter warnings:
5.1. CSP
5.2. Native Messaging
5.3. Content scripts for all URLs (not implemented yet).

1 out of every N consecutive updates for an add-on will be auto-approved, so a counter should be kept per-add-on. The counter shouldn’t change if a version is deleted by the developer, or if a new version is uploaded while the previous one was awaiting review (to avoid gaming via re-submission). The value N should be configurable, initially set to 5 and decreased over time if the outcome is positive. It can be set to 0 in order to shut down auto-approvals entirely.

Based on past performance, the initial setting should lead to ~40 auto-approved add-ons every week. It can be set to 0 in order to shut down auto-approvals.

The submission flow shouldn’t change. The last step should still tell the developer the add-on is pending review. Just like with unlisted add-ons, the review is performed automatically by the `Mozilla` user and a canned review message should be used both for the log and to send with the review email: `Thank you for your contribution.`. This is the same canned response used by reviewers when approving an add-on.

To ensure there aren’t serious security problems in the auto-approval system, the first implementation to go live should only identify the files that meet the auto-approval criteria and store it in the database, without auto-approving any files.

A report of identified files will be run manually so they can be verified by members of the security and review teams. Once the criteria is determined to be sufficient, auto-approvals can be turned on.

**Comment by kewisch** [comment]
I also think there should be a limit on the number of automatic reviews that can be done in succession. Even though the chance is slim, if there has been one automatic review I don't think there should be a second.

We should also anticipate developers that don't delete the version, but just continue uploading new versions with minimal changes just to get automatic review.

**Comment by jvillalobos** [comment]
I don't think it's necessarily bad to get a couple of consecutive of auto-approvals, and the global probability will determine if that's likely or not. It's not uncommon for a number of add-on updates to get rubber-stamped until someone discovers an issue that should have been brought up long before.

> continue uploading new versions with minimal changes

I included the case of the previous version already awaiting review. What else do you think should be done?

**Comment by andymckay** [comment]
It sounds like its just a random %, is that correct?

Wondering if we should aim for key APIs or permissions on an approved list. Eg:

* if you have zero permissions
* if you only have content_scripts with css in them
* if hosts only applies to one site

I did some preliminary examination of reviewed WebExtensions here:

https://gist.github.com/andymckay/10c3a4c64ce8990b589f0ac740f65955

But it would be easy enough to take some real scenarios and see how many extensions that would let through and hence the affect on the queues. If you just take extensions that have *no permissions*, that's **20.91%** of extensions (those extensions might define content scripts or something else, so let's not get too excited by that figure).

And yes, some things (eg: nativeMessaging) should drop people into an auto-review state.

**Comment by jvillalobos** [comment]
Yes, I'd like to start with a random %, given the urgency. There are many additions that we should do as followups, like the ones you mention, some metric for developer reputation, past review performance, etc.

**Comment by wagnerand** [comment]
I think it would be good to come up with some numbers first to measure the impact this will have. I think that we also shouldn't move forward without having some criteria that will make the review ineligible for automated review. One thing that concerns me is that in the past we had to learn the hard way that validation tools can be circumvented quite easily. Do we have a strategy for that, this time?
I am also concerned that the number of reviews we skip will quickly get out of hand and we will end up with a similar state like the countless numbers of unlisted versions we don't look at (despite the limited percentage).
Don't get me wrong, I'm not against this. But we had mitigation plans for unlisted too, and they have failed.

I agree with @kewisch that there shouldn't be too many consecutive auto-approvals. Another reason for that is that a manual review will be much harder and much likely to miss something the bigger the changeset (read: number of auto-approved versions) is.

I am wondering what the cost/benefit ratio of this is. If an add-on lands in the manual queue (for whatever reason) and a review is supposed to review all auto-approved versions since the last manual review (or the equivalent changeset), we are just deferring the problem to a later review and volunteers will either be even less likely to take big reviews (they rarely do already) or the reviews will have a lower quality (and the quality is already concerning as of now).

I'd like us to come up with a solid plan first instead of doing a quick implementation that hits us afterwards because we don't have time to follow up. This has happened in the past which did the opposite of what it was supposed to do: It increased workload for admin reviewers even more.

**Comment by jvillalobos** [comment]
> validation tools can be circumvented quite easily. Do we have a strategy for that, this time?

Circumventing validation will always be possible, no way around it. Note the initial proposal doesn't even take validation into account.

> But we had mitigation plans for unlisted too, and they have failed.

We don't have any mitigation plan for the current state of reviews either. What happens when the add-ons waiting for review are more than 1K. What happens when/if we lose our most active reviewer? We've been able to recover in the past, but I suspect we won't this time.

> there shouldn't be too many consecutive auto-approvals.

How many would you say is too many?

>It increased workload for admin reviewers even more.

Given the failure rate of regular human reviews, I don't think this will do worse. Of course, both of those can be difficult to quantify.

> I'd like us to come up with a solid plan first

Like I mentioned on IRC, I don't want to block on this. We've been discussing this for too long (years, depending on how you count). If this somehow turns out to be a terrible idea, we will set the % back to 0 and find a different approach or ways to improve on it, many of which have already been mentioned here.

**Comment by pauljt** [comment]
> Just like with unlisted add-ons, the review is performed automatically by the Mozilla user and a canned review message should be used both for the log and to send with the review email: Your submission has been chosen randomly to be automatically approved. This doesn't guarantee future automatic approvals.

Consider adding some lag here to prevent abuse. Also i dont see a reason to tell the developer any details of how the review of conducted, and telling them the review was automated is a bad idea. Why not just wait some arbitrary length of time (maybe varied even) then just send them the positive review signal as normal?

**Comment by pauljt** [comment]
> Yes, I'd like to start with a random %, given the urgency and busy pipeline. There are many additions that we should do as followups, like the ones you mention, some metric for developer reputation, past review performance, etc.

I'm pretty strongly opposed to using random as a heuristic here. Surely we can do better than that. 

> I'd like us to come up with a solid plan first instead of doing a quick implementation that hits us afterwards because we don't have time to follow up. This has happened in the past which did the opposite of what it was supposed to do: It increased workload for admin reviewers even more.

I'd second that approach.

**Comment by kewisch** [comment]
> Consider adding some lag here to prevent abuse. Also i dont see a reason to tell the developer any details of how the review of conducted, and telling them the review was automated is a bad idea. Why not just wait some arbitrary length of time (maybe varied even) then just send them the positive review signal as normal?

Review emails always contain a reviewer name. Automated reviews use "Mozilla", so people will catch wind of this pretty quickly. There may be other measures we can take that should be discussed privately.

> I'm pretty strongly opposed to using random as a heuristic here. Surely we can do better than that.

We can do better, but it will take more time to implement. I'm all for making some restrictions as noted above if they are easy to implement, but this is meant as a temporary™ measure to give us some room to breath while we work on a better solution. Right now we are spending a large chunk of the week only on reviews.

> > there shouldn't be too many consecutive auto-approvals.

> How many would you say is too many?
Can we configure this? I'd say start with 2 being too many.

**Comment by kewisch** [comment]
Another possible restriction: only auto-review if the last version was an approval (no deletes, rejections, or unreviewed versions).

**Comment by jvillalobos** [comment]
> Surely we can do better than that.

The current situation is untenable and we need a quick mitigation for the growing review queues. Some context:

* The review queues are in a very bad state and [getting worse](https://docs.google.com/spreadsheets/d/1txL7fXXohjW1SIB1x9Mata3_dXnwSZMx_CwzsGPCW7I/edit#gid=0).
* The admin reviewers are working non-stop trying to keep up.
* The AMO Engineering team already has a full plate for Q1, and this can't wait for Q2.
* This is limited to updates, meaning at least the initial submission has been reviewed by a human. Only a fraction of updates would be auto-approved.
* There's no way forward that doesn't involve automation and increased risk over the current setup.

> Why not just wait some arbitrary length of time (maybe varied even) then just send them the positive review signal as normal?

I don't think we can keep automated reviews a secret, especially considering as we will move more towards automation in the future. But I agree that saying it was done `randomly` is disclosing too much. I'll correct that.

**Comment by andymckay** [comment]
Is the code to implement a random amount of WebExtensions much harder than using some basic checks? Or just going and clicking randomly in the review tools?

**Comment by jvillalobos** [comment]
> Is the code to implement a random amount of WebExtensions much harder than using some basic checks?

Depends on the checks, of course, especially if they need to be introduced to the linter. Adding a check that doesn't depend on the add-on validation or metadata should be easier, but I could be wrong. It also makes it less flexible, in that this may need to be ramped up/down depending on the state of the queues.

> Or just going and clicking randomly in the review tools?

That takes people's time, and puts their name next to what's essentially a rubber stamp. I don't think that's a productive approach.

**Comment by kewisch** [comment]
Using some basic checks without randomness means relying fully on those basic checks for reviews? I don't think our current checks are advanced enough to do that. 

I don't know how much effort the implementation is, but clicking randomly in the review tools would work for a while. I guess I could do some ad-hoc automation using command line tools to support that. I don't know if this is a good idea for 5 months though. Also, in that case the review is done in our name, which may confuse developers if there are questions on a future rejection.

**Comment by kewisch** [comment]
> I am wondering what the cost/benefit ratio of this is. If an add-on lands in the manual queue (for whatever reason) and a reviewer is supposed to review all auto-approved versions since the last manual review (or the equivalent changeset), we are just deferring the problem to a later review

I would argue that reviewers (even admin) do not have to look at results of automatic reviews, but if they catch something immediately adjacent to a part of the diff they are reviewing, they should flag it (as they should anyway). This won't catch all issues, but that is the risk we are taking.

**Comment by muffinresearch** [comment]
> I'd like us to come up with a solid plan first instead of doing a quick implementation that hits us afterwards because we don't have time to follow up. This has happened in the past which did the opposite of what it was supposed to do: It increased workload for admin reviewers even more.

I too would against rushing to implement something that's not targetted in some way without giving some time to consider alternatives. 

If we take a more considered approach based on what we know about specific types of extensions (e.g. exactly the sort of thing @andymckay is suggesting) then at least we can qualify the risk we are taking and the manual review time can be spent where it's needed.

Ideally we need to consider all approaches to the problem so we can then find the best balance around time to implement vs risk management and potential longer term implications.

**Comment by andymckay** [comment]
A rather silly, but not too complicated way to get a risk assessment by just looking at the manifest: mozilla/addons-linter#1112. Also its pretty cool the stuff we get with WebExtensions.

**Comment by jvillalobos** [comment]
I've updated the initial comment with the current state of the proposal, after further discussions. I also started a [PRD for the larger plan](https://docs.google.com/document/d/1VxvvzQQvfDOVCylSVlpfeXv0tDYRLZfIMvWOoraANGU/edit#), which at the moment only includes a bit more context of this first phase and some notes on what we want to achieve for the followup in Q2.

**Comment by eviljeff** [comment]
@jvillalobos what about rejections?  Does having a version rejected not alter the chance of the next version being auto-approved?

**Comment by wagnerand** [comment]
I think we agreed that a submission following a rejection should not be eligible for auto-approval.

**Comment by atsay** [comment]
@eviljeff We clarified in the PRD

**Comment by jvillalobos** [comment]
Updated the initial comment to incorporate the previous approval requirement and content scripts for all sites as a restriction.

**Comment by pauljt** [comment]
Whats the status here? Did this end up happening?

**Comment by diox** [comment]
A prototype command has been merged. It does not auto-approve yet, but records what it would have done if it did. We ran it last Friday, now we need to tweak its rules according to the results. @jvillalobos for more info.

**Comment by jvillalobos** [comment]
@pauljt I sent an update on this last Friday. Let me know if you didn't get that email.

**Comment by diox** [comment]
What's left for this issue:
- Implement actual auto approval in the command
- Run it one more time manually (once in dry-run, once for real)
- Verify that everything went right, and run it via crontab

Once that is done I'll work on [post-review](https://docs.google.com/document/d/1hVx-NbuVRc0wKXSdt8eHceObexY93YFzMXtqYC95nUs/edit#) next - adding a list of auto approved add-ons/versions in reviewer tools, relaxing the criteria for auto approval step by step until all webextensions are auto approved - but that should be another issue.

**Comment by jvillalobos** [comment]
I reviewed the add-ons flagged by the first command run and nothing stood out as a problem. We're good to go with the current set of rules.

**Comment by ValentinaPC** [comment]
This can be marked as verified fixed because all its dependencies were fixed.
